File: go/server/choices_test.go
package server

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Unit tests for the choices Rich Interaction kind — the Go port of the Rust
// rust/smooth-operator/src/choices.rs tests, plus validation against the shared
// conformance fixtures (spec/conformance/fixtures.json), so the Go validator agrees
// with the Rust reference on the same wire shapes.

func opt(label string) ChoiceOption { return ChoiceOption{Label: label} }

func question(header string, labels []string, multi bool) ChoiceQuestion {
	opts := make([]ChoiceOption, len(labels))
	for i, l := range labels {
		opts[i] = opt(l)
	}
	return ChoiceQuestion{Question: header + "?", Header: header, Options: opts, MultiSelect: multi}
}

func answer(header string, options []string, other string) ChoiceAnswer {
	return ChoiceAnswer{Header: header, Options: options, Other: other}
}

func TestValidateChoicesValidSingleSelectNormalizes(t *testing.T) {
	qs := []ChoiceQuestion{question("Plan", []string{"Basic", "Pro"}, false)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", []string{"  Pro  "}, "")}}
	out, errs := validateChoices(qs, vals)
	if errs != nil {
		t.Fatalf("expected valid, got errors: %v", errs)
	}
	if len(out.Answers) != 1 || len(out.Answers[0].Options) != 1 || out.Answers[0].Options[0] != "Pro" {
		t.Fatalf("expected normalized [Pro], got %+v", out.Answers)
	}
	if out.Answers[0].Other != "" {
		t.Fatalf("expected no other, got %q", out.Answers[0].Other)
	}
}

func TestValidateChoicesMultiSelectKeepsAllPicks(t *testing.T) {
	qs := []ChoiceQuestion{question("Topics", []string{"Sales", "Support", "Billing"}, true)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Topics", []string{"Sales", "Billing"}, "")}}
	out, errs := validateChoices(qs, vals)
	if errs != nil {
		t.Fatalf("expected valid, got errors: %v", errs)
	}
	if strings.Join(out.Answers[0].Options, ",") != "Sales,Billing" {
		t.Fatalf("expected [Sales Billing], got %v", out.Answers[0].Options)
	}
}

func TestValidateChoicesOtherEscapeHatchAccepted(t *testing.T) {
	qs := []ChoiceQuestion{question("Plan", []string{"Basic", "Pro"}, false)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", nil, "  Enterprise, actually  ")}}
	out, errs := validateChoices(qs, vals)
	if errs != nil {
		t.Fatalf("expected valid, got errors: %v", errs)
	}
	if len(out.Answers[0].Options) != 0 {
		t.Fatalf("expected no option picks, got %v", out.Answers[0].Options)
	}
	if out.Answers[0].Other != "Enterprise, actually" {
		t.Fatalf("expected trimmed other, got %q", out.Answers[0].Other)
	}
}

func TestValidateChoicesUnknownLabelIsFieldError(t *testing.T) {
	qs := []ChoiceQuestion{question("Plan", []string{"Basic", "Pro"}, false)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", []string{"Platinum"}, "")}}
	_, errs := validateChoices(qs, vals)
	if len(errs) != 1 {
		t.Fatalf("expected 1 error, got %v", errs)
	}
	if errs[0].Field != "Plan" || !strings.Contains(errs[0].Message, "not one of the offered") {
		t.Fatalf("unexpected error: %+v", errs[0])
	}
}

func TestValidateChoicesSingleSelectRejectsMultiplePicks(t *testing.T) {
	qs := []ChoiceQuestion{question("Plan", []string{"Basic", "Pro"}, false)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", []string{"Basic", "Pro"}, "")}}
	_, errs := validateChoices(qs, vals)
	if !anyErrContains(errs, "single answer") {
		t.Fatalf("expected single-answer error, got %v", errs)
	}
}

func TestValidateChoicesUnansweredQuestionRequired(t *testing.T) {
	qs := []ChoiceQuestion{
		question("Plan", []string{"Basic", "Pro"}, false),
		question("Size", []string{"S", "M"}, false),
	}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", []string{"Pro"}, "")}}
	_, errs := validateChoices(qs, vals)
	if len(errs) != 1 || errs[0].Field != "Size" || !strings.Contains(errs[0].Message, "must be answered") {
		t.Fatalf("expected Size must-be-answered, got %v", errs)
	}
}

func TestValidateChoicesEmptyAnswerNeedsPickOrOther(t *testing.T) {
	qs := []ChoiceQuestion{question("Plan", []string{"Basic", "Pro"}, false)}
	vals := ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", nil, "")}}
	_, errs := validateChoices(qs, vals)
	if !anyErrContains(errs, "select an option") {
		t.Fatalf("expected select-an-option error, got %v", errs)
	}
}

func TestValidateChoicesFormatOnlyWhenSpecGone(t *testing.T) {
	// No questions (prior-turn fallback raise) → format-only: a pick is accepted as-is.
	out, errs := validateChoices(nil, ChoiceValues{Answers: []ChoiceAnswer{answer("Plan", []string{"Anything"}, "")}})
	if errs != nil {
		t.Fatalf("format-only should accept any pick, got %v", errs)
	}
	if out.Answers[0].Options[0] != "Anything" {
		t.Fatalf("format-only should keep the pick, got %v", out.Answers)
	}
	// …but an empty answer set still errors.
	_, errs = validateChoices(nil, ChoiceValues{})
	if !anyErrContains(errs, "provide an answer") {
		t.Fatalf("expected empty-answers error, got %v", errs)
	}
}

func TestParseQuestionsEnforcesContract(t *testing.T) {
	// Happy path with shorthand string options.
	qs, err := parseQuestions([]any{
		map[string]any{"question": "Which plan?", "header": "Plan", "options": []any{"Basic", "Pro"}},
	})
	if err != nil {
		t.Fatalf("valid parse failed: %v", err)
	}
	if len(qs) != 1 || qs[0].Options[0].Label != "Basic" || qs[0].MultiSelect {
		t.Fatalf("unexpected parse result: %+v", qs)
	}

	// Too many questions.
	tooMany := make([]any, 5)
	for i := range tooMany {
		tooMany[i] = map[string]any{"question": "q", "header": string(rune('A' + i)), "options": []any{"a", "b"}}
	}
	if _, err := parseQuestions(tooMany); err == nil {
		t.Fatal("expected error for >4 questions")
	}
	// Too few options.
	if _, err := parseQuestions([]any{map[string]any{"question": "q", "header": "H", "options": []any{"only"}}}); err == nil {
		t.Fatal("expected error for <2 options")
	}
	// Header too long.
	if _, err := parseQuestions([]any{map[string]any{"question": "q", "header": "ThisHeaderIsWayTooLong", "options": []any{"a", "b"}}}); err == nil {
		t.Fatal("expected error for long header")
	}
	// Duplicate headers.
	if _, err := parseQuestions([]any{
		map[string]any{"question": "q1", "header": "H", "options": []any{"a", "b"}},
		map[string]any{"question": "q2", "header": "H", "options": []any{"a", "b"}},
	}); err == nil {
		t.Fatal("expected error for duplicate headers")
	}
}

func TestChoicesKindWiresReferenceSurface(t *testing.T) {
	k := ChoicesKind{}
	if k.Kind() != "choices" || k.Capability() != "choice_chips" || k.ToolSchema().Name != "request_choices" {
		t.Fatalf("unexpected kind surface: %s/%s/%s", k.Kind(), k.Capability(), k.ToolSchema().Name)
	}

	req, err := k.ParseRequest(map[string]any{
		"questions": []any{map[string]any{"question": "Which plan interests you?", "header": "Plan",
			"options": []any{map[string]any{"label": "Basic"}, map[string]any{"label": "Pro"}}}},
		"reason": "to route you",
	})
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if req.Kind != "choices" || req.Reason != "to route you" {
		t.Fatalf("unexpected request: %+v", req)
	}
	var spec choicesSpec
	if err := json.Unmarshal(req.Spec, &spec); err != nil || spec.Questions[0].Header != "Plan" {
		t.Fatalf("unexpected spec: %s (%v)", req.Spec, err)
	}

	// The validator, through the kind, produces the canonical values.
	canonical, errs := k.Validate(req.Spec, map[string]any{"answers": []any{map[string]any{"header": "Plan", "options": []any{"Pro"}}}})
	if errs != nil {
		t.Fatalf("expected valid submit, got %v", errs)
	}
	cv := canonical.(ChoiceValues)
	if cv.Answers[0].Options[0] != "Pro" {
		t.Fatalf("unexpected canonical values: %+v", cv)
	}

	// The fallback directive enumerates the options.
	directive := k.FallbackDirective(req.Spec, "to route you")
	if !strings.Contains(directive, "Basic, Pro") {
		t.Fatalf("fallback directive missing enumerated options: %s", directive)
	}
}

// TestChoicesFixturesValidate cross-checks the Go kind against the shared conformance
// fixtures the Rust reference validates against: the choices_values submitted against the
// choices_spec must validate to the choices_payload's normalized answers.
func TestChoicesFixturesValidate(t *testing.T) {
	fixtures := loadFixtures(t)
	spec := fixtureInstance(t, fixtures, "choices_spec")
	values := fixtureInstance(t, fixtures, "choices_values")
	payload := fixtureInstance(t, fixtures, "choices_payload")

	specJSON, err := json.Marshal(spec)
	if err != nil {
		t.Fatalf("marshal spec: %v", err)
	}

	canonical, errs := ChoicesKind{}.Validate(specJSON, values)
	if errs != nil {
		t.Fatalf("shared choices_values should validate against choices_spec, got %v", errs)
	}

	// The canonical result must match the fixture payload's `values` (jsonEqual
	// normalizes both through JSON, so the struct tags line up with the wire shape).
	if !jsonEqual(canonical, payload["values"]) {
		got, _ := json.Marshal(canonical)
		want, _ := json.Marshal(payload["values"])
		t.Fatalf("canonical values != fixture payload values\n got: %s\nwant: %s", got, want)
	}
	if payload["status"] != "submitted" {
		t.Fatalf("fixture payload status = %v, want submitted", payload["status"])
	}
}

func loadFixtures(t *testing.T) map[string]any {
	t.Helper()
	path := filepath.Join("..", "..", "spec", "conformance", "fixtures.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixtures: %v", err)
	}
	var f map[string]any
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("parse fixtures: %v", err)
	}
	return f
}

func fixtureInstance(t *testing.T, fixtures map[string]any, key string) map[string]any {
	t.Helper()
	entry, ok := fixtures[key].(map[string]any)
	if !ok {
		t.Fatalf("fixture %q missing", key)
	}
	inst, ok := entry["instance"].(map[string]any)
	if !ok {
		t.Fatalf("fixture %q has no instance", key)
	}
	return inst
}

func anyErrContains(errs []InteractionFieldError, sub string) bool {
	for _, e := range errs {
		if strings.Contains(e.Message, sub) {
			return true
		}
	}
	return false
}

File: typescript/server/test/choices.test.ts
/**
 * The `choices` Rich Interaction kind — validator + kind wiring + the shared
 * conformance fixtures.
 *
 * The TS parity of the Rust `choices.rs` unit tests: the validator's one-pass
 * per-question rules, the raise-tool argument contract (`parseQuestions`), the
 * kind's `parseRequest`/`validate`/`fallbackDirective`, and the `interaction_required`
 * protocol builder — all cross-checked against the three shared `choices` fixtures
 * in `spec/conformance/fixtures.json` (so a drift in any one server is caught here).
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { describe, expect, it } from 'vitest';

import { ChoicesKind, HEADER_MAX_CHARS, parseQuestions, validateChoices, type ChoiceQuestion, type ChoiceValues } from '../src/choices.js';
import * as protocol from '../src/protocol.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SPEC_DIR = join(__dirname, '..', '..', '..', 'spec');
const fixtures = JSON.parse(readFileSync(join(SPEC_DIR, 'conformance', 'fixtures.json'), 'utf8')) as Record<string, { instance: Record<string, unknown> } | string>;

function fixture(name: string): Record<string, unknown> {
    const entry = fixtures[name];
    if (!entry || typeof entry === 'string') throw new Error(`missing fixture ${name}`);
    return entry.instance;
}

function q(header: string, labels: string[], multiSelect = false): ChoiceQuestion {
    return { question: `${header}?`, header, options: labels.map((l) => ({ label: l, description: '' })), multiSelect };
}
function vals(answers: ChoiceValues['answers']): ChoiceValues {
    return { answers };
}

describe('validateChoices', () => {
    it('single-select normalizes (trims labels)', () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: ['  Pro  '] }]));
        expect(out.ok).toBe(true);
        if (out.ok) {
            expect(out.values.answers).toHaveLength(1);
            expect(out.values.answers[0]!.options).toEqual(['Pro']);
            expect(out.values.answers[0]!.other).toBeUndefined();
        }
    });

    it('multi-select keeps all picks', () => {
        const out = validateChoices([q('Topics', ['Sales', 'Support', 'Billing'], true)], vals([{ header: 'Topics', options: ['Sales', 'Billing'] }]));
        expect(out.ok).toBe(true);
        if (out.ok) expect(out.values.answers[0]!.options).toEqual(['Sales', 'Billing']);
    });

    it("accepts the free-text 'Other' escape hatch and trims it; blank Other dropped", () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: [], other: '  Enterprise, actually  ' }]));
        expect(out.ok).toBe(true);
        if (out.ok) {
            expect(out.values.answers[0]!.options).toEqual([]);
            expect(out.values.answers[0]!.other).toBe('Enterprise, actually');
        }
        const blank = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: ['Pro'], other: '   ' }]));
        expect(blank.ok).toBe(true);
        if (blank.ok) expect(blank.values.answers[0]!.other).toBeUndefined();
    });

    it('unknown label is a per-question error', () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: ['Platinum'] }]));
        expect(out.ok).toBe(false);
        if (!out.ok) {
            expect(out.errors).toHaveLength(1);
            expect(out.errors[0]!.field).toBe('Plan');
            expect(out.errors[0]!.message).toContain('not one of the offered');
        }
    });

    it('single-select rejects multiple picks', () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: ['Basic', 'Pro'] }]));
        expect(out.ok).toBe(false);
        if (!out.ok) expect(out.errors.some((e) => e.message.includes('single answer'))).toBe(true);
    });

    it('every question must be answered', () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro']), q('Size', ['S', 'M'])], vals([{ header: 'Plan', options: ['Pro'] }]));
        expect(out.ok).toBe(false);
        if (!out.ok) {
            expect(out.errors).toHaveLength(1);
            expect(out.errors[0]!.field).toBe('Size');
            expect(out.errors[0]!.message).toContain('must be answered');
        }
    });

    it('an empty answer needs a pick or an other', () => {
        const out = validateChoices([q('Plan', ['Basic', 'Pro'])], vals([{ header: 'Plan', options: [] }]));
        expect(out.ok).toBe(false);
        if (!out.ok) expect(out.errors.some((e) => e.message.includes('select an option'))).toBe(true);
    });

    it('format-only path (no questions) accepts any answer with a pick', () => {
        const out = validateChoices([], vals([{ header: 'Anything', options: ['whatever'] }]));
        expect(out.ok).toBe(true);
        const empty = validateChoices([], vals([]));
        expect(empty.ok).toBe(false);
    });
});

describe('parseQuestions — the raise-tool contract', () => {
    it('accepts shorthand string options and defaults multiSelect to false', () => {
        const qs = parseQuestions([{ question: 'Which plan?', header: 'Plan', options: ['Basic', 'Pro'] }]);
        expect(qs).toHaveLength(1);
        expect(qs[0]!.options[0]!.label).toBe('Basic');
        expect(qs[0]!.multiSelect).toBe(false);
    });

    it('rejects >4 questions, <2 options, an over-long header, and duplicate headers', () => {
        expect(() => parseQuestions(Array.from({ length: 5 }, (_, i) => ({ question: 'q', header: `H${i}`, options: ['a', 'b'] })))).toThrow(/between 1 and 4/);
        expect(() => parseQuestions([{ question: 'q', header: 'H', options: ['only'] }])).toThrow(/between 2 and 4/);
        expect(() => parseQuestions([{ question: 'q', header: 'ThisHeaderIsWayTooLong', options: ['a', 'b'] }])).toThrow(/too long/);
        expect(() => parseQuestions([{ question: 'q1', header: 'H', options: ['a', 'b'] }, { question: 'q2', header: 'H', options: ['a', 'b'] }])).toThrow(/duplicate/);
    });

    it(`caps the header at ${HEADER_MAX_CHARS} characters`, () => {
        expect(() => parseQuestions([{ question: 'q', header: 'x'.repeat(HEADER_MAX_CHARS + 1), options: ['a', 'b'] }])).toThrow(/too long/);
        expect(parseQuestions([{ question: 'q', header: 'x'.repeat(HEADER_MAX_CHARS), options: ['a', 'b'] }])).toHaveLength(1);
    });
});

describe('ChoicesKind', () => {
    it('exposes the reference identity + raise-tool surface', () => {
        const kind = new ChoicesKind();
        expect(kind.kind).toBe('choices');
        expect(kind.capability).toBe('choice_chips');
        expect(kind.toolSchema().name).toBe('request_choices');
    });

    it('parseRequest canonicalizes questions + reason', () => {
        const req = new ChoicesKind().parseRequest({
            questions: [{ question: 'Which plan interests you?', header: 'Plan', options: [{ label: 'Basic' }, { label: 'Pro' }] }],
            reason: 'to route you',
        });
        expect(req.kind).toBe('choices');
        expect(req.reason).toBe('to route you');
        expect((req.spec as { questions: ChoiceQuestion[] }).questions[0]!.header).toBe('Plan');
    });

    it('fallbackDirective enumerates the options and points at submit_interaction', () => {
        const kind = new ChoicesKind();
        const req = kind.parseRequest({ questions: [{ question: 'Which plan?', header: 'Plan', options: ['Basic', 'Pro'] }], reason: 'to route you' });
        const directive = kind.fallbackDirective(req.spec, 'to route you');
        expect(directive).toContain('Basic, Pro');
        expect(directive).toContain('submit_interaction');
    });
});

describe('the shared choices fixtures', () => {
    it('validate(choices_spec, choices_values) produces choices_payload.values', () => {
        const spec = fixture('choices_spec');
        const values = fixture('choices_values');
        const payload = fixture('choices_payload');
        const out = new ChoicesKind().validate(spec, values);
        expect(out.ok).toBe(true);
        if (out.ok) expect(out.values).toEqual((payload as { values: unknown }).values);
    });

    it('interactionRequired builder matches the interaction_required_event shape, carrying the choices spec', () => {
        const spec = fixture('choices_spec');
        const built = protocol.interactionRequired('req-a1b2c3d4-0004', '88888888-8888-8888-8888-888888888888', 'choices', spec, 'to route you');
        const rt = JSON.parse(JSON.stringify(built)) as Record<string, unknown>;
        expect(rt.type).toBe('interaction_required');
        expect(rt.requestId).toBe('req-a1b2c3d4-0004');
        const inner = (rt.data as { data: Record<string, unknown> }).data;
        expect(inner.interactionId).toBe('88888888-8888-8888-8888-888888888888');
        expect(inner.kind).toBe('choices');
        expect(inner.spec).toEqual(spec);
        expect(inner.reason).toBe('to route you');
    });

    it('interactionInvalid builder carries per-field errors and stays retryable-shaped', () => {
        const built = protocol.interactionInvalid('req-1', 'i-1', 'choices', [{ field: 'Plan', message: 'this question must be answered' }], 'Some fields need attention.');
        const inner = (built.data as { data: Record<string, unknown> }).data;
        expect(built.type).toBe('interaction_invalid');
        expect((inner.errors as unknown[])).toHaveLength(1);
        expect((inner.errors as { field: string }[])[0]!.field).toBe('Plan');
        expect(inner.message).toBe('Some fields need attention.');
    });
});
